### PR TITLE
Firebase Integration to the App for storing the modules time

### PR DESCRIPTION
**Pull request by RickyRiko305**
|PR checklist|Status|
|-----|:------:|
|Is code analysis task executed and issues resolved in the Java/Kotlin code?|Yes|
|Is code analysis task executed and issues resolved in the CPP/native code?|NA|
|Is the version code set in the AndroidManifest.xml or build.gradle file to 1 more than the version which has already been published?|NA|
|Is code commenting done properly for the changes made?|Yes|
|Is correct variable naming convention followed?|Yes|
|Is auto formatting shortcut Ctrl+Alt+Shift+L executed in all files?|No|
|Are the message strings used in the application taken from strings.xml file?|Yes|
|Are the permissions which are not required removed from AndroidManifest.java file?|NA|
|Is the _uses-feature_ tag used for only must have feature in AndroidManifest.java file?|NA|
|Are the generated files correctly configured in .gitignore?|NA|
|Are the release signing configurations removed from the gradle files?|NA|
|Is commented code removed from all files?|Yes|
|Are updated images compressed (< 100 KB)?|NA|
|Ensured there are no typos in any string. For example: Variable/method names, hardcoded strings, commit messages, PR titles, issues, code comments?|Yes|

### CODE REVIEW

File: .idea/misc.xml [hunk 1]
```diff
@@ -35,7 +35,7 @@
       </value>
     </option>
   </component>
-  <component name="ProjectRootManager" version="2" languageLevel="JDK_1_7" default="true" project-jdk-name="1.8" project-jdk-type="JavaSDK">
+  <component name="ProjectRootManager" version="2" languageLevel="JDK_1_8" default="true" project-jdk-name="1.8" project-jdk-type="JavaSDK">
```
Comment (akashdabhi03) [inline .idea/misc.xml]: Revert this change

File: app/build.gradle [hunk 2]
```diff
@@ -56,6 +57,8 @@ android {
         implementation fileTree(dir: 'libs', include: ['*.jar'])
         testImplementation 'junit:junit:4.12'
         implementation 'com.android.support:appcompat-v7:28.0.0'
+        //28.0.0 'androidx.appcompat:appcompat:1.0.0-alpha3'//
+        //implementation 'com.google.firebase:firebase-core:17.0.0'
```
Comment (akashdabhi03) [inline app/build.gradle]: Remove commented code

File: app/build.gradle [hunk 3]
```diff
@@ -113,4 +116,8 @@ dependencies {
     implementation project(path: ':zip_file')
     implementation project(path: ':downloader_library')
     implementation 'com.android.support.constraint:constraint-layout:1.1.3'
-}
\ No newline at end of file
+    implementation 'com.google.firebase:firebase-database:16.0.4'
+    //implementation 'com.google.firebase:firebase-core:17.0.0'
```
Comment (akashdabhi03) [inline app/build.gradle]: Remove commented code

File: app/google-services.json [hunk 4]
```diff
@@ -0,0 +1,48 @@
+{
```
Comment (akashdabhi03) [inline app/google-services.json]: Remove this file

File: app/src/main/java/com/maq/xprize/onecourse/hindi/mainui/DownloadExpansionFile.java [hunk 5]
```diff
@@ -56,14 +56,14 @@
...
-                    1297005281L // the length of the patch file in bytes
-            )
+                    1198433694L//1544045005L // the length of the file in bytes
+            )//,
+//            new XAPKFile(
+//                    false, // false signifies a patch file
+//                    4, // the version of the APK that the patch file was uploaded
+//                    // against
+//                    1297005281L // the length of the patch file in bytes
+//            )
```
Comment (akashdabhi03) [inline app/src/main/java/com/maq/xprize/onecourse/hindi/mainui/DownloadExpansionFile.java]: Revert these changes

File: app/src/main/java/com/maq/xprize/onecourse/hindi/mainui/MainActivity.java [hunk 6]
```diff
@@ -137,9 +144,28 @@ public static OBGroup armPointer() {
         return arm;
     }
 
+    public static void logEvent(/*String id,*/String name,long st, long et/*, long elpT*/,String status){
```
Comment (akashdabhi03) [inline app/src/main/java/com/maq/xprize/onecourse/hindi/mainui/MainActivity.java]: Use better names:
name -> moduleName
st -> moduleStartTime
et -> moduleEndTime
status -> moduleStatus

File: app/src/main/java/com/maq/xprize/onecourse/hindi/mainui/MainActivity.java [hunk 7]
```diff
@@ -137,9 +144,28 @@ public static OBGroup armPointer() {
...
+        String s = name;
+        long elpT;
```
Comment (akashdabhi03) [inline app/src/main/java/com/maq/xprize/onecourse/hindi/mainui/MainActivity.java]: Rename to moduleElapsedTime

File: app/src/main/java/com/maq/xprize/onecourse/hindi/mainui/MainActivity.java [hunk 8]
```diff
@@ -137,9 +144,28 @@ public static OBGroup armPointer() {
...
+        int module_index = s.lastIndexOf("/");
```
Comment (akashdabhi03) [inline app/src/main/java/com/maq/xprize/onecourse/hindi/mainui/MainActivity.java]: Use camel case for variable names

Comment (akashdabhi03) [reply app/src/main/java/com/maq/xprize/onecourse/hindi/mainui/MainActivity.java]: Evaluate and change where ever applicable

File: app/src/main/java/com/maq/xprize/onecourse/hindi/mainui/OBMainViewController.java [hunk 9]
```diff
@@ -42,6 +44,11 @@
     private Integer currentTouchID;
     private Handler lowlightButtonHandler = new Handler();
 
+    private String lastModuleName ;
+    private long StartTime;
+    private long EndTime;
+    public static int userIID;
```
Comment (akashdabhi03) [inline app/src/main/java/com/maq/xprize/onecourse/hindi/mainui/OBMainViewController.java]: Remove userid and the related code

File: app/src/main/java/com/maq/xprize/onecourse/hindi/mainui/OBMainViewController.java [hunk 10]
```diff
@@ -288,6 +295,14 @@ public OBGLView glView ()
 
     public boolean glMode ()
     {
+        long t = 28829384;
```
Comment (akashdabhi03) [inline app/src/main/java/com/maq/xprize/onecourse/hindi/mainui/OBMainViewController.java]: No need of this since you are initializing the variable in the next line

File: app/src/main/java/com/maq/xprize/onecourse/hindi/mainui/OBMainViewController.java [hunk 11]
```diff
@@ -288,6 +295,14 @@ public OBGLView glView ()
...
+        t = System.currentTimeMillis()/1000;                                                                          //timeStamp in seconds.
+
+        if(lastModuleName != null ){
+            EndTime = t;
```
Comment (akashdabhi03) [inline app/src/main/java/com/maq/xprize/onecourse/hindi/mainui/OBMainViewController.java]: Remove this since it is not being used

File: app/src/main/java/com/maq/xprize/onecourse/hindi/mainui/OBMainViewController.java [hunk 12]
```diff
@@ -443,18 +460,58 @@ public void pushViewController (Class<?> vcClass, Boolean animate, boolean fromR
...
             @Override
             public void run ()
             {
+
+
+                String value = vc.getClass().getName();                                                                     // getting the name of the class
+
+                String moduleNamePath = value.replace("com.maq.xprize.onecourse.hindi.mainui.","");      // getting the name of the module which need to be loaded
+                boolean isFound = moduleNamePath.contains("OCM_Child");                                                    // checking the module if contain the OCM_child which is in oc_community. Used at the very start for menu.
+                boolean isFound1 = moduleNamePath.contains("OC_PlayZone");
```
Comment (akashdabhi03) [inline app/src/main/java/com/maq/xprize/onecourse/hindi/mainui/OBMainViewController.java]: isFound -> isCommunity
isFound1 -> isPlayZone

File: app/src/main/java/com/maq/xprize/onecourse/hindi/mainui/OBMainViewController.java [hunk 13]
```diff
@@ -443,18 +460,58 @@ public void pushViewController (Class<?> vcClass, Boolean animate, boolean fromR
...
+
+
+
                 vc.start();
             }
         });
+
+    }
+
+    public static void getuserID(int userID){                                                                             //get the userID;
```
Comment (akashdabhi03) [inline app/src/main/java/com/maq/xprize/onecourse/hindi/mainui/OBMainViewController.java]: Remove this function

File: app/src/main/java/com/maq/xprize/onecourse/hindi/utils/OC_FatController.java [hunk 14]
```diff
@@ -202,6 +204,18 @@ public void loadUserDB(DBSQL db)
 
     }
 
+    public void loadUserDBSQL(DBSQL db)
```
Comment (akashdabhi03) [inline app/src/main/java/com/maq/xprize/onecourse/hindi/utils/OC_FatController.java]: We can remove this for now

File: app/src/main/java/com/maq/xprize/onecourse/hindi/utils/User.java [hunk 15]
```diff
@@ -0,0 +1,74 @@
+package com.maq.xprize.onecourse.hindi.utils;
+
+public class User {
+
+    private String UserName;
+    private int Age;
```
Comment (akashdabhi03) [inline app/src/main/java/com/maq/xprize/onecourse/hindi/utils/User.java]: Remove username and age

File: app/src/main/java/com/maq/xprize/onecourse/hindi/utils/User.java [hunk 16]
```diff
@@ -0,0 +1,74 @@
+package com.maq.xprize.onecourse.hindi.utils;
```
Comment (akashdabhi03) [inline app/src/main/java/com/maq/xprize/onecourse/hindi/utils/User.java]: I don't think we are using this class. We can remove it.

File: market_licensing/build.gradle [hunk 17]
```diff
@@ -1,4 +1,5 @@
 apply plugin: 'com.android.library'
+//apply plugin: 'com.google.gms.google-services'
```
Comment (akashdabhi03) [inline market_licensing/build.gradle]: Remove

File: market_licensing/build.gradle [hunk 18]
```diff
@@ -9,10 +10,18 @@ android {
...
         release {
             minifyEnabled false
             proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.txt'
         }
     }
 }
+
+//dependencies {
+ //   implementation 'com.google.firebase:firebase-database:16.0.4'
+//}
```
Comment (akashdabhi03) [inline market_licensing/build.gradle]: Remove

File: zip_file/build.gradle [hunk 19]
```diff
@@ -1,4 +1,5 @@
 apply plugin: 'com.android.library'
+//apply plugin: 'com.google.gms.google-services'
```
Comment (akashdabhi03) [inline zip_file/build.gradle]: Remove

File: zip_file/build.gradle [hunk 20]
```diff
@@ -9,10 +10,18 @@ android {
...
+
     buildTypes {
         release {
             minifyEnabled false
             proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.txt'
         }
     }
 }
+
+//dependencies {
```
Comment (akashdabhi03) [inline zip_file/build.gradle]: Remove

**Review by akashdabhi03 — CHANGES_REQUESTED** [review]

File: app/google-services.json (hunk 4 above)
Comment (akashdabhi03) [inline app/google-services.json]: This file is still not removed

File: app/src/main/java/com/maq/xprize/onecourse/hindi/mainui/MainActivity.java [hunk 21]
```diff
@@ -117,7 +119,8 @@
     public ReentrantLock suspendLock = new ReentrantLock();
     float sfxMasterVolume = 1.0f;
     Map<String, Float> sfxVolumes = new HashMap<>();
-    private int b;
+    private static FirebaseAnalytics FirebaseAnalytics;
```
Comment (akashdabhi03) [inline app/src/main/java/com/maq/xprize/onecourse/hindi/mainui/MainActivity.java]: Camel case variables

File: app/src/main/java/com/maq/xprize/onecourse/hindi/mainui/OBMainViewController.java [hunk 22]
```diff
@@ -562,86 +525,87 @@ public void popViewControllerZoom (RectF zoomRect)
...
-    {
-        if(viewControllers != null && viewControllers.size() > 0)
-        {
-            OBSectionController controller = viewControllers.get(viewControllers.size()-1);
-            if(controller != null)
+    public void onResume() {
+        if (viewControllers != null && viewControllers.size() > 0) {
+            OBSectionController controller = viewControllers.get(viewControllers.size() - 1);
+            long endTime;
+            endTime = System.currentTimeMillis() / 1000;                                                                 // timeStamps in seconds
```
Comment (akashdabhi03) [inline app/src/main/java/com/maq/xprize/onecourse/hindi/mainui/OBMainViewController.java]: The name endTime is deceptive. We are anyways storing the value as startTime. Shouldn't it be startTime instead?

File: app/src/main/java/com/maq/xprize/onecourse/hindi/mainui/OBMainViewController.java [hunk 23]
```diff
@@ -42,8 +41,11 @@
     private Integer currentTouchID;
     private Handler lowlightButtonHandler = new Handler();
 
-    public OBMainViewController (Activity a)
-    {
+    private String lastModuleName;
+    private long StartTime;
```
Comment (akashdabhi03) [inline app/src/main/java/com/maq/xprize/onecourse/hindi/mainui/OBMainViewController.java]: startTime

**Review by akashdabhi03 — CHANGES_REQUESTED** [review]

File: app/google-services.json (hunk 4 above)
Comment (RickyRiko305) [reply app/google-services.json]: that file is needed for the Firebase Analytics

File: app/src/main/java/com/maq/xprize/onecourse/hindi/mainui/OBMainViewController.java [hunk 24]
```diff
@@ -388,6 +400,36 @@ public void pushViewController(Class<?> vcClass, Boolean animate, boolean fromRi
...
+                boolean isCommunity = moduleNamePath.contains("OCM_Child");                                                    // checking the module if contain the OCM_child which is in oc_community. Used at the very start for menu.
+                boolean isPlayZone = moduleNamePath.contains("OC_PlayZone");
+
+                String moduleName = moduleNamePath.replace(".", "/");
+                if (!isCommunity && !isPlayZone) {                                                                                              // we don't need to load the OCM_child, it's a menu
+                    long endTime;
+                    endTime = System.currentTimeMillis() / 1000;                                                                   // timeStamp in seconds
+
+                    if (lastModuleName != null) {
+                        MainActivity.logEvent(lastModuleName, startTime, endTime, "Completed");                             // calling the event log for ending the module.
```
Comment (akashdabhi03) [inline app/src/main/java/com/maq/xprize/onecourse/hindi/mainui/OBMainViewController.java]: Store Completed and In progress in Strings.xml file

**Review by akashdabhi03 — CHANGES_REQUESTED** [review]

**Review by akashdabhi03 — APPROVED** [review]

**Review by mathewnik90 — APPROVED** [review]
